File: data_compression/huffman.go
package main

import (
	"container/heap"
	"fmt"
)

type Node struct {
	char        rune
	freq        int
	left, right *Node
}

type PriorityQueue []*Node

func (pq PriorityQueue) Len() int { return len(pq) }

func (pq PriorityQueue) Less(i, j int) bool {
	return pq[i].freq < pq[j].freq
}

func (pq PriorityQueue) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
}

func (pq *PriorityQueue) Push(x interface{}) {
	*pq = append(*pq, x.(*Node))
}

func (pq *PriorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	*pq = old[0 : n-1]
	return item
}

func buildTree(text string) *Node {
	freqMap := make(map[rune]int)
	for _, c := range text {
		freqMap[c]++
	}

	pq := make(PriorityQueue, 0)
	heap.Init(&pq)

	for char, freq := range freqMap {
		heap.Push(&pq, &Node{char: char, freq: freq})
	}

	for pq.Len() > 1 {
		left := heap.Pop(&pq).(*Node)
		right := heap.Pop(&pq).(*Node)

		parent := &Node{
			char:  '$',
			freq:  left.freq + right.freq,
			left:  left,
			right: right,
		}

		heap.Push(&pq, parent)
	}

	if pq.Len() > 0 {
		return heap.Pop(&pq).(*Node)
	}
	return nil
}

func printCodes(root *Node, str string) {
	if root == nil {
		return
	}

	if root.char != '$' {
		fmt.Printf("%c: %s\n", root.char, str)
	}

	printCodes(root.left, str+"0")
	printCodes(root.right, str+"1")
}

func main() {
	text := "this is an example for huffman encoding"
	root := buildTree(text)
	printCodes(root, "")
}

File: data_compression/huffman.rs
/**
 * Huffman Coding in Rust
 */

use std::cmp::Ordering;
use std::collections::{BinaryHeap, HashMap};

#[derive(Debug, Eq, PartialEq)]
struct Node {
    freq: usize,
    char: Option<char>,
    left: Option<Box<Node>>,
    right: Option<Box<Node>>,
}

impl Ord for Node {
    fn cmp(&self, other: &Self) -> Ordering {
        other.freq.cmp(&self.freq)
    }
}

impl PartialOrd for Node {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

fn build_tree(text: &str) -> Option<Box<Node>> {
    let mut freq_map = HashMap::new();
    for c in text.chars() {
        *freq_map.entry(c).or_insert(0) += 1;
    }

    let mut heap = BinaryHeap::new();
    for (char, freq) in freq_map {
        heap.push(Box::new(Node {
            freq,
            char: Some(char),
            left: None,
            right: None,
        }));
    }

    while heap.len() > 1 {
        let left = heap.pop().unwrap();
        let right = heap.pop().unwrap();

        let parent = Box::new(Node {
            freq: left.freq + right.freq,
            char: None,
            left: Some(left),
            right: Some(right),
        });

        heap.push(parent);
    }

    heap.pop()
}

fn print_codes(node: &Option<Box<Node>>, str: String) {
    if let Some(n) = node {
        if let Some(c) = n.char {
            println!("{}: {}", c, str);
        } else {
            print_codes(&n.left, format!("{}0", str));
            print_codes(&n.right, format!("{}1", str));
        }
    }
}

fn main() {
    let text = "this is an example for huffman encoding";
    let root = build_tree(text);
    print_codes(&root, String::from(""));
}
